### ISSUE TITLE
Major Rewrite in Progress

### ISSUE THREAD

**Issue by RealJohnGalt** [Galts-Gulch/avarice#31]
I've decided to change the direction of Avarice a bit after being rather aimless with this project for quite some time.

After getting new threaded/locking API code half done, I decided to redo most of Avarice to be fully threaded. I'm also working on a gtk3 gui and a layer between gui, threads, config files, and logging to support new native gui easier. This will additionally allow multiple indicator instances to run side by side if used in a "preset" (described below).

Next I'll be redoing configobj support in the layer to support reading config presets. Any new indicator level configuration (combined configuration for trading) will be written to a preset in this path. At launch, you may select which created presets you which to run, and they'll be run separately.

I'm starting with the gtk3 gui (using pygobject) so I can see the progress imediately and be better motivated on the project.

To reiterate and summarize:
- basic (disconnected) gtk3 gui for motivation
- thread all the things
- add layer between GUI, threads, config files, and logging, to make support for new platform-specific gui support easier.
- rewrite configobj support in the new layer to handle indicator presets.
- Add actual functionality to gtk3 gui with new layer.
- Pythonwin (still undecided) and PyObjc windows/cocoa (osx) gui.

(after some release at this point):
- Batch trade configurables with similar preset support as indicators.

I still believe that with a decent interface, there are many traders who want something easy to configure with trade strategy presets, but don't want to give their API keys to a web trading platform.

**Comment by RealJohnGalt** [comment]
As a note: I considered using TK or Wx for the interface, but the problem is aesthetics. I believe it would be better to use a layer to bridge the gap between the GUI code, and use pygobject for gtk3 support in linux (get the pretty gnome/ubuntuisms). OSX users could get a cocoa interface using PyObjc, and Pythonwin (or a "less native" gui) could be used for windows.

Realistically due to userbase, it would go Linux pygobject => windows => osx. After step 4 in the summary above, it would be rather quick to sychronously work on windows and osx support however.

**Comment by ghost** [comment]
Sounds like a pretty ambitious project. I'm here and will help bug test if you need anything. I can't code for shit, but I'm pretty decent at finding glitches. :)

**Comment by RealJohnGalt** [comment]
Thank you, I always appreciate the help.

I have the base for the topmost bullet done. Onto the second and third for now since I'm a bit more motivated seeing something pretty :).

The HeaderBar "Start/Stop" button is a ToggleButton. It will start/stop main thread on toggle.

![wip gtk3 base](https://cloud.githubusercontent.com/assets/7730016/9513775/88a8b96c-4c48-11e5-83b6-850be411bb36.png)

**Comment by ghost** [comment]
Looks good. It's clean and functional. can't wait to try it.

**Comment by RealJohnGalt** [comment]
Ty @Hapexamendios.

**Update:**
- GUI layer partially written. This handles starting/stopping threads and serving/writing configuration.
- Some small backend changes completed for storing simulator and trade history.
- Made **storage** lock/release on shelve open/close. Added a new database configuration option for max history to store (defaults to 500 orders). **loggerdb** doesn't need to be thread safe (yet).
- Added a sortable/filterable grid to Trade History and Simulator History. This will read new simulatorhistory and tradehistory shelve files.

![Example of grid filled with fake data (didn't want to run)](https://cloud.githubusercontent.com/assets/7730016/9558897/12f2d870-4da0-11e5-8582-dc509c8eb432.png)

**Edit:** Just noticed we need to also store and include trade time.

**Comment by RealJohnGalt** [comment]
gui-prep branch for some of the backend work I've committed/merged. I'm holding off on the new (very wip) websocket api for now (though I might need to merge it in sooner due to some threading asyncio loop issues I've run into).

**Comment by RealJohnGalt** [comment]
start/stop functionality now working with a new process.
![screenshot from 2015-08-29 01-34-11](https://cloud.githubusercontent.com/assets/7730016/9561148/4b2a92de-4dee-11e5-825c-d3a19cd5ae22.png)

**Comment by RealJohnGalt** [comment]
I'm working on the Trade Indicator selection page now and attempting to make it as intuitive as possible. I'm also getting the gui layer setup for configurables using dictionaries and lists for serving options/descriptions.

The Trade Indicator selection page has all indicators listed down the left side of the page, with "combine" and "independent" checkboxes on the right. There are "New," "Save," and "Remove" buttons which add/remove at the top depending on configuration. You can select indicators to combine, then independent indicators, and click "Save," then start all over if "New" is clicked again. You can remove the previous configuration by clicking "Remove" next to the listed configuration.

I'm having a hard time coming up with a good name for this page. I don't believe that "Trade Indicators" is a very good name any longer. Any suggestions?

**Comment by ghost** [comment]
Hmmm Indicator index? :)

**Comment by RealJohnGalt** [comment]
Thank you @Hapexamendios .

I like it better than "Trade Indicators," but I don't think it describes the functionality for setting the indicators to be used. I just came up with a new idea though.

What do you think of "Indicator Strategy," or "Indicator Strategies?" Something like "Trade Strategy" seems to be most descriptive, but I don't want people to think it only applies to trader and not simulator. I might be over thinking this though :).

Anyway, I'm open to as many ideas as you can throw out. It will be a bit before this is completely ready for release, so names can change easily.

**Comment by RealJohnGalt** [comment]
What about just "Strategy" or "Strategies"?

**Comment by ghost** [comment]
I like strategies. :)

**Comment by RealJohnGalt** [comment]
It's decided :+1: 

Regarding the "rewrite," it's going to be rather minor at first. I'm focusing on getting the GUI layer done and integrated. From there I'll be working on threading, making indicators take arguments, and store in separate keys using arguments.  Former capabilities will be used for preset support to allow running multiple instances of an indicator with differing configurations.

I'd like to end up with a repo of preset strategies the GUI layer can pull from. If a user would like to share a strategy preset, they may do so with a pull request on the repo.

Preset support is going to be the absolute last to be completed. After preset support, windows and osx GUIs will be thrown together using GUI layer.

**Comment by RealJohnGalt** [comment]
Indicator configuration page and indicator configuration backend (guilayer) nearly complete (https://github.com/Galts-Gulch/avarice/commits/gui-prep). Strategies page up next, then general configuration page/backend.

**Comment by RealJohnGalt** [comment]
![screenshot from 2015-09-02 00-36-48](https://cloud.githubusercontent.com/assets/7730016/9625603/329fa65c-510c-11e5-9403-b19afa15145e.png)

Indicator page getting there. Needs some guilayer work for adjustment sizes for SpinButtons, layout work (divider underneath **Indicator** row for instance), ComboBox types added, buttons sorted in a more sane manner (like config.ini), and defaults loaded for each button type. All indicator pages are generated from guilayer to make the code more portable for different gui.

Once I have a good structure, the general configuration page will be quick to hammer out. The only difference will be a stack_switcher (like the history tab in previous images) for General Configuration. Due to the number of indicators, this wasn't a good option for Indicators.

Indicator page uses a hidden notebook for each indicator. indicator pages are stored in an auto scroll window. Changing the Indicator ComboBox changes the notebook page and ComboBox default on the selected page.

**Comment by RealJohnGalt** [comment]
I've been rather busy recently, but almost have the GUI and gui-prep complete :+1:

Edit: have a few new TODOs including "restart" prompts on configuration changes, gtk3 notifier support, and some threading improvements.

**Comment by aehaynes** [comment]
Good stuff!

**Comment by RealJohnGalt** [comment]
Thanks man.

New TODO:
- Restart prompts and gtk3 notifier support still needed.
- Progressbar in headerbar when more data is required (e.g. at initial run).
- Tutorial wizard.
- Improve svg graph rendering.
- Price buy/sell chart.
- Config argument on run, move config.ini to config.default.ini in preparation for presets.

After first release:
- Avarice API
- Daemon support, allowing for connecting to a server with the GUI.

Also, will probably reset versioning entirely and be saner about it. I don't want to give people the wrong idea about the project.

**Comment by RealJohnGalt** [comment]
Yes, this is still a WIP. I got held up with real work, trading, and getting a decent API figured out.

**Comment by ghostnegotiator** [comment]
This is definitely interesting, I am hoping I can contribute soon! :+1:
